airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/JobNotifier.java

```diff
@@ -0,0 +1,150 @@
+import io.airbyte.config.persistence.PersistenceConstants;
+import io.airbyte.notification.NotificationClient;
+import io.airbyte.notification.SlackNotificationClient;
+import io.airbyte.scheduler.models.Job;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.UUID;
+import org.apache.logging.log4j.util.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JobNotifier {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(JobNotifier.class);
+
+  private final String connectionPageUrl;
+  private final ConfigRepository configRepository;
+
+  public JobNotifier(String webappUrl, ConfigRepository configRepository) {
+    this.connectionPageUrl = String.format("%s/source/connection/", webappUrl);
+    this.configRepository = configRepository;
+  }
+
+  public void failJob(final String reason, final Job job) {
+    final UUID connectionId = UUID.fromString(job.getScope());
+    final UUID sourceDefinitionId = configRepository.getSourceDefinitionFromConnection(connectionId).getSourceDefinitionId();
+    final UUID destinationDefinitionId = configRepository.getDestinationDefinitionFromConnection(connectionId).getDestinationDefinitionId();
+    try {
+      final ImmutableMap<String, Object> sourceDefMetadata = generateSourceDefinitionMetadata(sourceDefinitionId);
+      final ImmutableMap<String, Object> destinationDefMetadata = generateDestinationDefinitionMetadata(destinationDefinitionId);
+      final ImmutableMap<String, Object> jobMetadata = generateJobMetadata(reason, job);
+      final Map<String, Object> metadata = MoreMaps.merge(sourceDefMetadata, destinationDefMetadata, jobMetadata);
+      final StandardWorkspace workspace = configRepository.getStandardWorkspace(PersistenceConstants.DEFAULT_WORKSPACE_ID, true);
+
+      for (Notification notification : workspace.getNotifications()) {
+        final NotificationClient notificationClient = getNotificationClient(notification);
+        try {
+          notificationClient.notifyJobFailure(metadata);
+        } catch (InterruptedException | IOException e) {
+          LOGGER.error("Failed to notify: {} due to {}", notification, e);
+        }
+      }
+    } catch (JsonValidationException | IOException | ConfigNotFoundException e) {
+      LOGGER.error("Unable to read Workspace configuration");
+    }
+  }
+
+  private NotificationClient getNotificationClient(final Notification notification) {
+    if (notification.getNotificationType() == NotificationType.SLACK) {
+      return new SlackNotificationClient(connectionPageUrl, notification.getSlackConfiguration());
+    } else {
+      throw new IllegalArgumentException("Unknown notification type:" + notification.getNotificationType());
+    }
+  }
+
+  private ImmutableMap<String, Object> generateDestinationDefinitionMetadata(UUID destinationDefinitionId)
```

Comment:
are some of these methods copy pastes from the JobTracker? If so is there a reasonable DRY them up?